pkg/image/api/helper.go

```diff
@@ -290,6 +371,30 @@ func DockerImageReferenceForStream(stream *ImageStream) (DockerImageReference, e
 	return ParseDockerImageReference(spec)
 }
 
+// FollowTagReference walks through the defined tags on a stream, following any referential tags in the stream.
+// Will return ok if the tag is valid, multiple if the tag had at least reference, and ref and finalTag will be the last
+// tag seen. If a circular loop is found ok will be false.
+func FollowTagReference(stream *ImageStream, tag string) (finalTag string, ref *TagReference, ok bool, multiple bool) {
```

Comment:
Much easier to follow. Only comment now is that four return values are no fun to think about, and although *some* callers of this (exported) method are written by you, the rest might not be, so you most likely want to give `len(seen.List()) > 1` as the `multiple` return always. 

+1 for not using the fact that `multiple` is initialized to false ...